# SingleCell
![](https://camo.githubusercontent.com/27d70bea7d28e1346bed05f5e2fc26ef79101832ef77de0aeca37d703d301be0/68747470733a2f2f646174612e77686963646e2e636f6d2f696d616765732f3139313132303537392f6f726967696e616c2e676966)
## tutorials
https://satijalab.org/seurat/index.html<br>
https://scanpy.readthedocs.io/en/stable/tutorials.html<br>
https://github.com/hbctraining/scRNA-seq_online/blob/master/schedule/links-to-lessons.md<br>
https://www.singlecellcourse.org/index.html<br>
https://bioconductor.org/books/3.15/OSCA.advanced<br>
## Doublet Analysis
Scrublet: https://github.com/swolock/scrublet<br>
DoubletFinder: https://github.com/chris-mcginnis-ucsf/DoubletFinder<br>
scDblFinder: https://github.com/plger/scDblFinder<br>
Benchmark article: [Benchmarking Computational Doublet-Detection Methods for Single-Cell RNA Sequencing Data](https://www.cell.com/cell-systems/fulltext/S2405-4712(20)30459-2?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS2405471220304592%3Fshowall%3Dtrue)

##  Single cell data integration
harmony: https://github.com/immunogenomics/harmony<br>
LIGER: https://github.com/welch-lab/liger<br>
sctransform: https://github.com/satijalab/sctransform<br>
scanorama： https://github.com/brianhie/scanorama<br>
Conos: https://github.com/kharchenkolab/conos<br>
Benchmark: [Benchmarking atlas-level data integration in single-cell genomics](https://www.nature.com/articles/s41592-021-01336-8)

## Single cell annotation
scCATCH: https://github.com/ZJUFanLab/scCATCH<br>
sc-type: https://github.com/IanevskiAleksandr/sc-type<br>
SingleR: https://github.com/LTLA/SingleR<br>
clustermole： https://github.com/igordot/clustermole<br>
UNIFAN: https://github.com/doraadong/UNIFAN<br>
MACA: https://github.com/ImXman/MACA<br>
DISCO: https://www.immunesinglecell.org/cellpredictor<br>
scAnnotatR: https://github.com/grisslab/scAnnotatR<br>
celltypist: https://github.com/Teichlab/celltypist<br>
scPred: https://github.com/powellgenomicslab/scPred<br>
Benchmark: https://github.com/tabdelaal/scRNAseq_Benchmark<br>

## Trajectory Inference
Monocle3: https://cole-trapnell-lab.github.io/monocle3<br>
Slingshot: https://bioconductor.org/packages/devel/bioc/vignettes/slingshot/inst/doc/vignette.html<br>
Palantir: https://github.com/dpeerlab/Palantir<br>
scSTEM: https://github.com/alexQiSong/scSTEM<br>
Tempora: https://github.com/BaderLab/Tempora<br>
Benchmark: [A comparison of single-cell trajectory inference methods](https://www.nature.com/articles/s41587-019-0071-9)<br>


## Single cell gene enrichment
escape: https://github.com/ncborcherding/escape<br>
ssGSEA2.0: https://github.com/broadinstitute/ssGSEA2.0<br>
GSVA：https://github.com/rcastelo/GSVA<br>
AUCell: https://github.com/aertslab/AUCell<br>
scGSVA: https://github.com/guokai8/scGSVA<br>
## Cell-Cell communication
CellChat: https://github.com/sqjin/CellChat<br>
CellPhoneDB: https://github.com/Teichlab/cellphonedb<br>
nichenetr: https://github.com/saeyslab/nichenetr<br>
celltalker：https://github.com/arc85/celltalker<br>
## RNA velocity
scVelo: https://github.com/theislab/scvelo<br>
dynamo: https://github.com/aristoteleo/dynamo-release<br>
## Single cell regulatory network
pySCENIC: https://github.com/aertslab/pySCENIC<br>
SIGNET: https://github.com/Lan-lab/SIGNET<br>
Benchmark article https://www.nature.com/articles/s41592-019-0690-6<br>
## Inference of CNV
InferCNV: https://github.com/broadinstitute/infercnv<br>
CopyKAT: https://github.com/navinlabcode/copykat<br>
HoneyBADGER: https://github.com/JEFworks-Lab/HoneyBADGER<br>
## Cell fate
scFates: https://github.com/LouisFaure/scFates<br>

## Tumour cells
ikarus: https://github.com/BIMSBbioinfo/ikarus<br>

## Visualization
scINSIGHT: https://github.com/Vivianstats/scINSIGHT<br>
scPubr: https://enblacar.github.io/SCpubr-book/index.html<br>

## Analyse pipeline
popsicleR: https://github.com/bicciatolab/popsicleR<br>
## Lab
https://github.com/ggjlab<br>
https://github.com/ZJUFanLab<br>
https://github.com/theislab<br>
## Tools colection
https://www.scrna-tools.org/<br>
